presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -401,11 +402,31 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
             Type fieldType = checkedTypeLookup(symbol);
             NullableValue value = normalized.getValue();
 
-            Optional<NullableValue> coercedValue = coerce(value, fieldType);
-            if (coercedValue.isPresent()) {
-                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+            if (!normalized.wasNonCoercionCastRemoved()) {
```

Comment:
This code is somewhat tricky.

Can we assume that left and right operand of `ComparisonExpression` passed to `visitCompareExpression` have same type?
Before this PR tests assumed that it is possible.

Cases like this were tested:
```
s = 1000
```

 where `s` is of type DECIMAL and 1000 is `INTEGER`.
Can that actually happen during execution.

What I observed is that I always got sth like:
```
s = CAST(1000 as DECIMAL)
```

One exception I observed is for null.
```
s = NULL
```
came without any casts to neither `UNKNOWN` or `DECIMAL`.

If we can make such assumption then condition 
```
!normalized.wasNonCoercionCastRemoved()
```
could be changed to:
```
!wasCastRemoved()
```

As cases that would got through the first one but not the second one would be filtered out anyway by:
```
if (coercedValue.isPresent())
```